src/orbit_determination/tooshortarc.jl

```diff
@@ -0,0 +1,718 @@
+    r = q[1:3] + ρ * ρ_unit + params.eph_su(t)[1:3]
+    # Barycentric velocity
+    v = q[4:6] + v_ρ * ρ_unit + ρ * v_α * ρ_α + ρ * v_δ * ρ_δ
+        + params.eph_su(t)[4:6]
+    # Barycentric state vector
+    return vcat(r, v)
+end
+
+# Propagate an orbit and compute residuals
+function propres(radec::Vector{RadecMPC{T}}, jd0::U, q0::Vector{V}, params::NEOParameters{T};
+                 dynamics::D = newtonian!)  where {D, T <: AbstractFloat, U <: Number, V <: Number}
+    # Time of first (last) observation
+    t0, tf = datetime2julian(date(radec[1])), datetime2julian(date(radec[end]))
+    # Epoch (plain)
+    _jd0_ = cte(cte(jd0))
+    # Years in backward (forward) integration
+    nyears_bwd = -(_jd0_ - t0 + params.bwdoffset) / yr
+    nyears_fwd = (tf - _jd0_ + params.fwdoffset) / yr
+    # Backward (forward) integration
+    bwd = propagate(dynamics, jd0, nyears_bwd, q0, params)
+    fwd = propagate(dynamics, jd0, nyears_fwd, q0, params)
+    if !issuccessfulprop(bwd, t0 - _jd0_; tol = params.coeffstol) ||
+       !issuccessfulprop(fwd, tf - _jd0_; tol = params.coeffstol)
+        return bwd, fwd, Vector{OpticalResidual{T, U}}(undef, 0)
+    end
+    # O-C residuals
+    res = residuals(radec, params;
+                    xvs = et -> auday2kmsec(params.eph_su(et/daysec)),
+                    xve = et -> auday2kmsec(params.eph_ea(et/daysec)),
+                    xva = et -> bwdfwdeph(et, bwd, fwd))
+
+    return bwd, fwd, res
+end
+
+@doc raw"""
+    adam(radec::Vector{RadecMPC{T}}, A::AdmissibleRegion{T}, ρ::T, v_ρ::T,
+         params::NEOParameters{T}; scale::Symbol = :linear, η::T = 25.0,
+         μ::T = 0.75, ν::T = 0.9, ϵ::T = 1e-8, Qtol::T = 0.001,
+         dynamics::D = newtonian!) where {T <: AbstractFloat, D}
+
+Adaptative moment estimation (ADAM) minimizer of normalized mean square
+residual over and admissible region `A`.
+
+!!! warning
+    This function will set the (global) `TaylorSeries` variables to `dx dy`.
+
+!!! reference
+    See Algorithm 1 of https://doi.org/10.48550/arXiv.1412.6980.
+"""
+function adam(radec::Vector{RadecMPC{T}}, A::AdmissibleRegion{T}, ρ::T, v_ρ::T,
+              params::NEOParameters{T}; scale::Symbol = :linear, η::T = 25.0,
+              μ::T = 0.75, ν::T = 0.9, ϵ::T = 1e-8, Qtol::T = 0.001,
+              dynamics::D = newtonian!) where {T <: AbstractFloat, D}
+    # Initial time of integration [julian days]
+    jd0 = datetime2julian(A.date)
+    # Nominal Line-Of-Sight
+    LOS = [A.α, A.δ, A.v_α, A.v_δ]
+    # Scaling factors
+    scalings = Vector{T}(undef, 6)
+    scalings[1:4] .= LOS ./ 1e6
+    if scale == :linear
+        scalings[5] = (A.ρ_domain[2] - A.ρ_domain[1]) / 1_000
+    elseif scale == :log
+        x = log10(ρ)
+        scalings[5] = (log10(A.ρ_domain[2]) - log10(A.ρ_domain[1])) / 1_000
+    end
+    scalings[6] = (A.v_ρ_domain[2] - A.v_ρ_domain[1]) / 1_000
+    # Jet transport variables
+    dq = scaled_variables("dx", scalings, order = 2)
```

Comment:
I'd suggest to keep `order` as a keyword argument to `adam`

Reply:
Solved by https://github.com/PerezHz/NEOs.jl/pull/31/commits/b67b85ec7f3e4b5d0bf443cce1388194d6fbf464.